### PR TITLE
Allow scale up/down in KRaft mode

### PR DESCRIPTION
**Pull request by showuon**
### Type of change

_Select the type of your PR_

- Bugfix

### Description

_Please describe your pull request_

Fix https://github.com/strimzi/strimzi-kafka-operator/issues/6862 , to allow scale up/down in KRaft mode by removing `quorum-state` file before launching kafka

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: systemtest/src/test/java/io/strimzi/systemtest/rollingupdate/RollingUpdateST.java [hunk 1]
```diff
@@ -245,8 +245,7 @@ void testRecoveryDuringKafkaRollingUpdate(ExtensionContext extensionContext) {
     @ParallelNamespaceTest
     @Tag(ACCEPTANCE)
     @Tag(COMPONENT_SCALING)
-    @KRaftNotSupported("Zookeeper is not supported by KRaft mode and is used in this test case")
-    void testKafkaAndZookeeperScaleUpScaleDown(ExtensionContext extensionContext) {
+    void testKafkaScaleUpScaleDown(ExtensionContext extensionContext) {
```
Comment (scholzj) [inline systemtest/src/test/java/io/strimzi/systemtest/rollingupdate/RollingUpdateST.java:248]: ZooKeeper is still there - it is only skipped when running in the KRaft mode. So maybe its not needed to rename the test?

Comment (showuon) [reply systemtest/src/test/java/io/strimzi/systemtest/rollingupdate/RollingUpdateST.java]: updated. Thanks.

**Comment by ppatierno** [comment]
@showuon I still see "WIP" in the title. Is it just a leftover and is it ready for review now?

**Comment by showuon** [comment]
> @showuon I still see "WIP" in the title. Is it just a leftover and is it ready for review now?

Sorry for the confusion, it's ready for review now. Thanks.

File: docker-images/kafka-based/kafka/scripts/kafka_run.sh [hunk 2]
```diff
@@ -63,6 +63,12 @@ if [ "$STRIMZI_KRAFT_ENABLED" = "true" ]; then
     echo "Kraft storage is already formatted"
   fi
 
+  # remove quorum-state file so that we won't enter voter not match error after scaling up/down
+  if [ -f "$KRAFT_LOG_DIR/__cluster_metadata-0/quorum-state" ]; then
+    echo "Removing quorum-state file"
+    rm -f "$KRAFT_LOG_DIR/__cluster_metadata-0/quorum-state"
+  fi
```
Comment (ppatierno) [inline docker-images/kafka-based/kafka/scripts/kafka_run.sh:70]: is it safe to remove quorum-state file even just when brokers are rolled for any other reason? (i.e. config change, updates, ...) So I mean, when it's not about scale up/down ...

Comment (showuon) [reply docker-images/kafka-based/kafka/scripts/kafka_run.sh:70]: I know what you mean. This is safe. The `quorum-state` file contains the voter's info and the leader's info. That means, Before this change, when a pod is restarted, the controller (suppose it is not the leader) can directly fetch from the quorum leader node. After this change, we have to find the current quorum leader each time the pod is restarted. I think that's the only difference. 


I think it should be fine. WDYT?

Comment (ppatierno) [reply docker-images/kafka-based/kafka/scripts/kafka_run.sh:70]: More rolling happens in specific scenarios (i.e. certificates renewal, ...) but it should not happen that frequently.
I guess this "fine the current quorum leader" is about some RPC calls.
I can live with it. @scholzj wdyt?

Comment (showuon) [reply docker-images/kafka-based/kafka/scripts/kafka_run.sh:70]: > I guess this "fine the current quorum leader" is about some RPC calls.

Correct

**Review by ppatierno — APPROVED** [review]
LGTM. Assuming all tests pass.

File: systemtest/src/test/java/io/strimzi/systemtest/AbstractST.java [hunk 3]
```diff
@@ -478,10 +478,12 @@ protected void testDockerImagesForKafkaCluster(String clusterName, String cluste
             kafkaVersion = Environment.ST_KAFKA_VERSION;
         }
 
-        //Verifying docker image for zookeeper pods
-        for (int i = 0; i < zkPods; i++) {
-            String imgFromPod = PodUtils.getContainerImageNameFromPod(kafkaNamespaceName, KafkaResources.zookeeperPodName(clusterName, i), "zookeeper");
-            assertThat("Zookeeper pod " + i + " uses wrong image", imgFromPod, containsString(TestUtils.parseImageMap(imgFromDeplConf.get(KAFKA_IMAGE_MAP)).get(kafkaVersion)));
+        if (!isKRaftEnabled) {
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/AbstractST.java:481]: I guess you can use the `Environment.isKRaftModeEnabled()` directly here and you don't have to pass it as a method parameter.

(Same applies for the second if)

**Comment by showuon** [comment]
For maintainers, we can re-run the system tests now. Thanks.

**Comment by im-konge** [comment]
Thanks @showuon ! :)

**Comment by showuon** [comment]
It failed again. I'll take a look tomorrow. Thanks.

**Comment by scholzj** [comment]
@im-konge Is that just the failure with broken test clients again? Can you please have a look so that Luke doesn't have to waste his time on unrelated issues?

**Comment by im-konge** [comment]
In the logs of the producer I can see:
```
2023-03-24 11:11:16.459.3905289416205 [33mWARN [m Sender:624 - [Producer clientId=producer-1] Got error produce response with correlation id 2086 on topic-partition my-topic-1632537724-466472546-0, retrying (2147482710 attempts left). Error: NOT_ENOUGH_REPLICAS
```
Which is not related to the issue with test-clients.

File: systemtest/src/test/java/io/strimzi/systemtest/rollingupdate/RollingUpdateST.java [hunk 4]
```diff
@@ -245,15 +245,15 @@ void testRecoveryDuringKafkaRollingUpdate(ExtensionContext extensionContext) {
...
     void testKafkaAndZookeeperScaleUpScaleDown(ExtensionContext extensionContext) {
         final TestStorage testStorage = new TestStorage(extensionContext, namespace);
 
         resourceManager.createResource(extensionContext, KafkaTemplates.kafkaPersistent(testStorage.getClusterName(), 3, 3)
             .editSpec()
                 .editKafka()
-                    .addToConfig(singletonMap("default.replication.factor", 1))
-                    .addToConfig("auto.create.topics.enable", "false")
+                    // Topic operator doesn't support KRaft, yet, using auto topic creation and default replication factor as workaround
+                    .addToConfig(singletonMap("default.replication.factor", Environment.isKRaftModeEnabled()? 3 : 1))
```
Comment (showuon) [inline systemtest/src/test/java/io/strimzi/systemtest/rollingupdate/RollingUpdateST.java:255]: This is the reason why it reported `NOT_ENOUGH_REPLICAS` error. Fixed.

File: systemtest/src/test/java/io/strimzi/systemtest/rollingupdate/RollingUpdateST.java [hunk 5]
```diff
@@ -341,7 +344,7 @@ void testKafkaAndZookeeperScaleUpScaleDown(ExtensionContext extensionContext) {
         assertThat(kubeClient().listPersistentVolumeClaims(testStorage.getNamespaceName(), testStorage.getClusterName()).stream()
```
Comment (showuon) [inline systemtest/src/test/java/io/strimzi/systemtest/rollingupdate/RollingUpdateST.java:344]: This assertion will fail in KRaft mode but pass in ZK mode. I need help here. 
In ZK mode, the pvc got deleted soon after Kafka pod is deleted. But in KRaft mode, the pvc will stay for ~2 mins after kafka pod deleted, and then deleted by itself. Do you know why this happened?
Should I look into it? Or workaround it by adding a helper waitPVCDeleted method to fix it? 
I'd like to ask for your advice first.

Thanks.

**Comment by scholzj** [comment]
Still seems to be failing with some issues (but a new one at least):

```
[ERROR] Failures: 
[ERROR] io.strimzi.systemtest.rollingupdate.RollingUpdateST.testKafkaAndZookeeperScaleUpScaleDown(ExtensionContext)
[ERROR]   Run 1: RollingUpdateST.testKafkaAndZookeeperScaleUpScaleDown:345 
Expected: is <3>
     but: was <7>
[ERROR]   Run 2: RollingUpdateST.testKafkaAndZookeeperScaleUpScaleDown:345 
Expected: is <3>
     but: was <7>
[ERROR]   Run 3: RollingUpdateST.testKafkaAndZookeeperScaleUpScaleDown:345 
Expected: is <3>
     but: was <7>
```

File: systemtest/src/test/java/io/strimzi/systemtest/rollingupdate/RollingUpdateST.java (hunk 5 above)
Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/rollingupdate/RollingUpdateST.java:345]: I think it is just a race condition when PVC waits to be garbage collected, `waitPVCDeleted` is a good idea from my POV. @scholzj is my assumption correct?

**Comment by Frawless** [comment]
/azp run kraft-regression

🤞

**Review by Frawless — APPROVED** [review]

**Review by scholzj — APPROVED** [review]
Thanks @showuon

**Comment by nupreti2022** [comment]
Today while working with client I had to delete the quorum-state file when increasing the number of controllers from 1 to 4. The CP version used is 7.4 in kraft mode. Please tell if the latest version of 7.4 has fixed the scaling issue.

**Comment by scholzj** [comment]
@nupreti2022 Assuming CP means Confluent Platform, you should probably direct your question at Confluent.

**Comment by zenvzenv** [comment]
Is this change specific to docker? I don't use docker, the version is 3.4.1, the problem still occurs
```text
Configured voter set: [1, 2] is different from the voter set read from the state file: [1]. Check if the quorum configuration is up to date, or wipe out the local state file if necessary
```
I added a new vote in config/kraft/server.properties, as follows:

Before adding:
```properties
controller.quorum.voters=1@10.20.74.57:9093
```
After adding:
```properties
controller.quorum.voters=1@10.20.74.57:9093,2@10.20.74.58:9093
```
Then I get an error when I start it again, how should I solve it?

**Comment by scholzj** [comment]
Sorry, what change are you talking about exactly? And what do you mean with _specific to Docker_? Strimzi runs on Kubernetes.

**Pull request merged**